In [1]:
import pandas as pd
#!conda install -c conda-forge folium=0.5.0 --yes
import folium
import requests
import json
from pandas.io.json import json_normalize
import numpy as np

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

## Use the Notebook to build the code to scrape the following Wikipedia page, https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M, in order to obtain the data that is in the table of postal codes and to transform the data into a pandas dataframe 

In [2]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
# scrap with pandas method read_html and import only table zero
df = pd.read_html(url)[0]

## Only process the cells that have an assigned borough. Ignore cells with a borough that is Not assigned.

In [3]:
df1=df.loc[df['Borough'] != 'Not assigned']

In [4]:
df1.head()

Postcode           Borough     Neighbourhood
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront
5      M6A        North York  Lawrence Heights
6      M6A        North York    Lawrence Manor

##  when there is "Not assigned" neighborhoods, then the neighborhood will be the same as the borough

In [5]:
df1.loc[df1['Neighbourhood'] =='Not assigned', 'Neighbourhood'] = df1['Borough']

/opt/conda/envs/Python36/lib/python3.6/site-packages/pandas/core/indexing.py:190: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self._setitem_with_indexer(indexer, value)
/opt/conda/envs/Python36/lib/python3.6/site-packages/ipykernel/__main__.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if __name__ == '__main__':


## Post code rows with same neighborhood will be combined into one row with the neighborhoods separated with a comma.

In [6]:
# group df1 by column "Postcode" & "borough" to concatenate Neibourhood with comma+space separator
# create new dataframe index 
df2 = df1.groupby(['Postcode','Borough']).agg( ', '.join).reset_index()
df2.head(10)

Postcode      Borough                                    Neighbourhood
0      M1B  Scarborough                                   Rouge, Malvern
1      M1C  Scarborough           Highland Creek, Rouge Hill, Port Union
2      M1E  Scarborough                Guildwood, Morningside, West Hill
3      M1G  Scarborough                                           Woburn
4      M1H  Scarborough                                        Cedarbrae
5      M1J  Scarborough                              Scarborough Village
6      M1K  Scarborough      East Birchmount Park, Ionview, Kennedy Park
7      M1L  Scarborough                  Clairlea, Golden Mile, Oakridge
8      M1M  Scarborough  Cliffcrest, Cliffside, Scarborough Village West
9      M1N  Scarborough                      Birch Cliff, Cliffside West

# In the last cell of your notebook, use the .shape method to print the number of rows of your dataframe

In [7]:
df2.shape

(103, 3)

### 103 rows & 3 Columns

# Use the Geocoder csv file to create a dataframe & add Latitude, Longitude Column to the dataframe

In [8]:
# read csv file and create dataframe
LatLong=pd.read_csv('http://cocl.us/Geospatial_data')
# rename df column Postal code to Postcode
LatLong.columns=['Postcode','Latitude','Longitude']
# merge dataframe LatLong with df2 on "Postcode" column
Toronto_Data= pd.merge(df2,LatLong[['Postcode','Latitude', 'Longitude']],on='Postcode')

### let's look at the Toronto coord by Neighbourhood

In [9]:
Toronto_Data.head()

Postcode      Borough                           Neighbourhood   Latitude  \
0      M1B  Scarborough                          Rouge, Malvern  43.806686   
1      M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2      M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3      M1G  Scarborough                                  Woburn  43.770992   
4      M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

In [10]:
Toronto_Data.shape

(103, 5)

## Surimpose Toronto Neighbourhoods with circle marks in a map

In [11]:
Map_Toronto = folium.Map(location=[43.660719, -79.376603],zoom_start=10)


for lat,lng,borough,neighbourhood in zip(Toronto_Data['Latitude'],Toronto_Data['Longitude'],Toronto_Data['Borough'],Toronto_Data['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
    [lat,lng],
    radius=5,
    popup=label,
    color='blue',
    fill=True,
    fill_color='#3186cc',
    fill_opacity=0.7,
    parse_html=False).add_to(Map_Toronto)
Map_Toronto

## Gett the top 100 venues of each Neighbourhood within a radius of 500 meters.

### Foursquare credential & url

In [12]:
CLIENT_ID = 'EWGDRUUA50BJPHG5W0ENE0GGWYY3A5MNHSEV5Z3OJBAZHJM4' # your Foursquare ID
CLIENT_SECRET = 'YW3KLXWUUHSUJDVRE4ANZRAUWMTYRZSLQBZ23D2GZBML2YUC' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

In [13]:
neighborhood_latitude = Toronto_Data.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = Toronto_Data.loc[0, 'Longitude'] # neighborhood longitude value
neighborhood_name = Toronto_Data.loc[0, 'Neighbourhood'] # neighborhood name

In [14]:
LIMIT = 100
radius = 500
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url 

'https://api.foursquare.com/v2/venues/explore?&client_id=EWGDRUUA50BJPHG5W0ENE0GGWYY3A5MNHSEV5Z3OJBAZHJM4&client_secret=YW3KLXWUUHSUJDVRE4ANZRAUWMTYRZSLQBZ23D2GZBML2YUC&v=20180605&ll=43.806686299999996,-79.19435340000001&radius=500&limit=100'

### Define a function to get the venues data

In [15]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

### Store venues in a dataframe

In [16]:
Toronto_venues = getNearbyVenues(names=Toronto_Data['Neighbourhood'],
                                   latitudes=Toronto_Data['Latitude'],
                                   longitudes=Toronto_Data['Longitude']
                                  )

Rouge, Malvern
Highland Creek, Rouge Hill, Port Union
Guildwood, Morningside, West Hill
Woburn
Cedarbrae
Scarborough Village
East Birchmount Park, Ionview, Kennedy Park
Clairlea, Golden Mile, Oakridge
Cliffcrest, Cliffside, Scarborough Village West
Birch Cliff, Cliffside West
Dorset Park, Scarborough Town Centre, Wexford Heights
Maryvale, Wexford
Agincourt
Clarks Corners, Sullivan, Tam O'Shanter
Agincourt North, L'Amoreaux East, Milliken, Steeles East
L'Amoreaux West
Upper Rouge
Hillcrest Village
Fairview, Henry Farm, Oriole
Bayview Village
Silver Hills, York Mills
Newtonbrook, Willowdale
Willowdale South
York Mills West
Willowdale West
Parkwoods
Don Mills North
Flemingdon Park, Don Mills South
Bathurst Manor, Downsview North, Wilson Heights
Northwood Park, York University
CFB Toronto, Downsview East
Downsview West
Downsview Central
Downsview Northwest
Victoria Village
Woodbine Gardens, Parkview Hill
Woodbine Heights
The Beaches
Leaside
Thorncliffe Park
East Toronto
The Danforth West, 

### explore dataframe

In [17]:
Toronto_venues.head(10)

Neighbourhood  Neighbourhood Latitude  \
0                          Rouge, Malvern               43.806686   
1  Highland Creek, Rouge Hill, Port Union               43.784535   
2  Highland Creek, Rouge Hill, Port Union               43.784535   
3       Guildwood, Morningside, West Hill               43.763573   
4       Guildwood, Morningside, West Hill               43.763573   
5       Guildwood, Morningside, West Hill               43.763573   
6       Guildwood, Morningside, West Hill               43.763573   
7       Guildwood, Morningside, West Hill               43.763573   
8       Guildwood, Morningside, West Hill               43.763573   
9       Guildwood, Morningside, West Hill               43.763573   

   Neighbourhood Longitude                            Venue  Venue Latitude  \
0               -79.194353                          Wendy's       43.807448   
1               -79.160497           Chris Effects Painting       43.784343   
2               -79.160497            Royal Canadian Legion       43.782533   
3               -79.188711  Swiss Chalet Rotisserie & Grill       43.767697   
4               -79.188711                G & G Electronics       43.765309   
5               -79.188711                 Big Bite Burrito       43.766299   
6               -79.188711            Enterprise Rent-A-Car       43.764076   
7               -79.188711            Woburn Medical Centre       43.766631   
8               -79.188711     Lawrence Ave E & Kingston Rd       43.767704   
9               -79.188711                         Eggsmart       43.767800   

   Venue Longitude              Venue Category  
0       -79.199056        Fast Food Restaurant  
1       -79.163742  Construction & Landscaping  
2       -79.163085                         Bar  
3       -79.189914                 Pizza Place  
4       -79.191537           Electronics Store  
5       -79.190720          Mexican Restaurant  
6       -79.193406         Rental Car Location  
7       -79.192286              Medical Center  
8       -79.189490                Intersection  
9       -79.190466              Breakfast Spot

In [18]:
Toronto_venues.shape

(2217, 7)

In [19]:
print('There are {} uniques categories.'.format(len(Toronto_venues['Venue Category'].unique())))

There are 270 uniques categories.


### 2217 venues returned by Foursquare 270 venues category

### Let's look how many venues per Neighbourhood

In [20]:
Toronto_venues.groupby('Neighbourhood').count()

Neighbourhood Latitude  \
Neighbourhood                                                                
Adelaide, King, Richmond                                               100   
Agincourt                                                                5   
Agincourt North, L'Amoreaux East, Milliken, Ste...                       3   
Albion Gardens, Beaumond Heights, Humbergate, J...                      11   
Alderwood, Long Branch                                                   9   
Bathurst Manor, Downsview North, Wilson Heights                         19   
Bayview Village                                                          4   
Bedford Park, Lawrence Manor East                                       26   
Berczy Park                                                             55   
Birch Cliff, Cliffside West                                              4   
Bloordale Gardens, Eringate, Markland Wood, Old...                       9   
Brockton, Exhibition Place, Parkdale Village                            25   
Business Reply Mail Processing Centre 969 Eastern                       15   
CFB Toronto, Downsview East                                              3   
CN Tower, Bathurst Quay, Island airport, Harbou...                      18   
Cabbagetown, St. James Town                                             45   
Caledonia-Fairbanks                                                      4   
Canada Post Gateway Processing Centre                                   10   
Cedarbrae                                                                8   
Central Bay Street                                                      79   
Chinatown, Grange Park, Kensington Market                               86   
Christie                                                                19   
Church and Wellesley                                                    82   
Clairlea, Golden Mile, Oakridge                                         10   
Clarks Corners, Sullivan, Tam O'Shanter                                 11   
Cliffcrest, Cliffside, Scarborough Village West                          3   
Cloverdale, Islington, Martin Grove, Princess G...                       2   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              37   
Davisville North                                                         7   
...                                                                    ...   
Northwood Park, York University                                          7   
Parkdale, Roncesvalles                                                  13   
Parkwoods                                                                2   
Queen's Park                                                            40   
Rosedale                                                                 4   
Roselawn                                                                 1   
Rouge, Malvern                                                           1   
Runnymede, Swansea                                                      38   
Ryerson, Garden District                                               100   
Scarborough Village                                                      1   
Silver Hills, York Mills                                                 1   
St. James Town                                                         100   
Stn A PO Boxes 25 The Esplanade                                         94   
Studio District                                                         42   
The Annex, North Midtown, Yorkville                                     22   
The Beaches                                                              4   
The Beaches West, India Bazaar                                          18   
The Danforth West, Riverdale                                            42   
The Junction North, Runnymede                                            4

### Analyze each Neighbourhood and input '1' when there is a venue and '0' when there is not a venue

In [21]:
# one hot encoding
Toronto_onehot = pd.get_dummies(Toronto_venues[['Venue Category']], prefix="", prefix_sep="")
# add neighborhood column back to dataframe
Toronto_onehot['Neighbourhood'] = Toronto_venues['Neighbourhood']
# move neighborhood column to the first column
fixed_columns = [Toronto_onehot.columns[-1]] + list(Toronto_onehot.columns[:-1])
Toronto_onehot = Toronto_onehot[fixed_columns]
Toronto_onehot.head()


Neighbourhood  Accessories Store  \
0                          Rouge, Malvern                  0   
1  Highland Creek, Rouge Hill, Port Union                  0   
2  Highland Creek, Rouge Hill, Port Union                  0   
3       Guildwood, Morningside, West Hill                  0   
4       Guildwood, Morningside, West Hill                  0   

   Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                  0        0                   0             0   
1                  0        0                   0             0   
2                  0        0                   0             0   
3                  0        0                   0             0   
4                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   ...  Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0  ...                              0                 0            0   
1  ...                              0                 0            0   
2  ...                              0                 0            0   
3  ...                              0                 0            0   
4  ...                              0                 0            0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wine Shop  Wings Joint  \
0                      0                0         0          0            0   
1                      0                0         0          0            0   
2                      0                0         0          0            0   
3                      0                0         0          0            0   
4                      0                0         0          0            0   

   Women's Store  Yoga Studio  
0              0            0  
1              0            0  
2              0            0  
3              0            0  
4              0            0  

[5 rows x 271 columns]

In [22]:
Toronto_onehot.shape

(2217, 271)

### let's get the frequency of venues for each Neighbourhood

In [23]:
Toronto_grouped = Toronto_onehot.groupby(['Neighbourhood']).mean().reset_index()

Toronto_grouped

Neighbourhood  Accessories Store  \
0                             Adelaide, King, Richmond                0.0   
1                                            Agincourt                0.0   
2    Agincourt North, L'Amoreaux East, Milliken, St...                0.0   
3    Albion Gardens, Beaumond Heights, Humbergate, ...                0.0   
4                               Alderwood, Long Branch                0.0   
5      Bathurst Manor, Downsview North, Wilson Heights                0.0   
6                                      Bayview Village                0.0   
7                    Bedford Park, Lawrence Manor East                0.0   
8                                          Berczy Park                0.0   
9                          Birch Cliff, Cliffside West                0.0   
10   Bloordale Gardens, Eringate, Markland Wood, Ol...                0.0   
11        Brockton, Exhibition Place, Parkdale Village                0.0   
12   Business Reply Mail Processing Centre 969 Eastern                0.0   
13                         CFB Toronto, Downsview East                0.0   
14   CN Tower, Bathurst Quay, Island airport, Harbo...                0.0   
15                         Cabbagetown, St. James Town                0.0   
16                                 Caledonia-Fairbanks                0.0   
17               Canada Post Gateway Processing Centre                0.0   
18                                           Cedarbrae                0.0   
19                                  Central Bay Street                0.0   
20           Chinatown, Grange Park, Kensington Market                0.0   
21                                            Christie                0.0   
22                                Church and Wellesley                0.0   
23                     Clairlea, Golden Mile, Oakridge                0.0   
24             Clarks Corners, Sullivan, Tam O'Shanter                0.0   
25     Cliffcrest, Cliffside, Scarborough Village West                0.0   
26   Cloverdale, Islington, Martin Grove, Princess ...                0.0   
27                      Commerce Court, Victoria Hotel                0.0   
28                                          Davisville                0.0   
29                                    Davisville North                0.0   
..                                                 ...                ...   
71                     Northwood Park, York University                0.0   
72                              Parkdale, Roncesvalles                0.0   
73                                           Parkwoods                0.0   
74                                        Queen's Park                0.0   
75                                            Rosedale                0.0   
76                                            Roselawn                0.0   
77                                      Rouge, Malvern                0.0   
78                                  Runnymede, Swansea                0.0   
79                            Ryerson, Garden District                0.0   
80                                 Scarborough Village                0.0   
81                            Silver Hills, York Mills                0.0   
82                                      St. James Town                0.0   
83                     Stn A PO Boxes 25 The Esplanade                0.0   
84                                     Studio District                0.0   
85                 The Annex, North Midtown, Yorkville                0.0   
86                                         The Beaches                0.0   
87                      The Beaches West, India Bazaar                0.0   
88                        The Danforth West, Riverdale                0.0   
89                       The Junction North, Runnymede                0.0   
90       The Kingsway, Montgomery Road, Old Mill North                0.0   
91                                   

In [24]:
num_top_venues = 5

for hood in Toronto_grouped['Neighbourhood']:
    print("----"+hood+"----")
    temp = Toronto_grouped[Toronto_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide, King, Richmond----
         venue  freq
0  Coffee Shop  0.07
1         Café  0.04
2   Steakhouse  0.04
3       Bakery  0.03
4          Bar  0.03


----Agincourt----
                       venue  freq
0                     Lounge   0.2
1             Breakfast Spot   0.2
2  Latin American Restaurant   0.2
3             Clothing Store   0.2
4               Skating Rink   0.2


----Agincourt North, L'Amoreaux East, Milliken, Steeles East----
               venue  freq
0         Playground  0.33
1               Park  0.33
2   Sculpture Garden  0.33
3  Accessories Store  0.00
4      Metro Station  0.00


----Albion Gardens, Beaumond Heights, Humbergate, Jamestown, Mount Olive, Silverstone, South Steeles, Thistletown----
            venue  freq
0   Grocery Store  0.18
1     Pizza Place  0.09
2  Discount Store  0.09
3        Pharmacy  0.09
4      Beer Store  0.09


----Alderwood, Long Branch----
          venue  freq
0   Pizza Place  0.22
1           Gym  0.11
2  Skating Rink  0.

### let's define a function that returns the x most common venues

In [25]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

### build a dataframe with the 10 x most frequent venues for each Neighbourhood

In [26]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = Toronto_grouped['Neighbourhood']

for ind in np.arange(Toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(Toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood  \
0                           Adelaide, King, Richmond   
1                                          Agincourt   
2  Agincourt North, L'Amoreaux East, Milliken, St...   
3  Albion Gardens, Beaumond Heights, Humbergate, ...   
4                             Alderwood, Long Branch   

       1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0                Coffee Shop            Steakhouse                  Café   
1  Latin American Restaurant                Lounge        Clothing Store   
2                       Park            Playground      Sculpture Garden   
3              Grocery Store            Beer Store   Fried Chicken Joint   
4                Pizza Place           Coffee Shop                  Pool   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0       Thai Restaurant                   Bar        Cosmetics Shop   
1          Skating Rink        Breakfast Spot           Yoga Studio   
2           Yoga Studio        Discount Store         Deli / Bodega   
3   Japanese Restaurant  Fast Food Restaurant        Discount Store   
4          Skating Rink                   Gym              Pharmacy   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0                Bakery      Asian Restaurant            Restaurant   
1      Doner Restaurant    Dim Sum Restaurant                 Diner   
2      Department Store          Dessert Shop    Dim Sum Restaurant   
3           Coffee Shop           Pizza Place        Sandwich Place   
4                   Pub        Sandwich Place          Dance Studio   

          10th Most Common Venue  
0  Vegetarian / Vegan Restaurant  
1                 Discount Store  
2                          Diner  
3                       Pharmacy  
4                  Deli / Bodega

## Cluster Neighborhoods in 5 clusters

In [27]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = Toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 2, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

### Let's create a new dataframe that includes the cluster for each neighborhood.

In [28]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)
neighborhoods_venues_sorted.head()
Toronto_merged = Toronto_Data

### Let's include the Post code & Borough for each neighborhood.

In [29]:
# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
Toronto_merged= pd.merge(Toronto_merged,neighborhoods_venues_sorted,on='Neighbourhood')
#Toronto_merged = Toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

Toronto_merged.head() # check the last columns!

Postcode      Borough                           Neighbourhood   Latitude  \
0      M1B  Scarborough                          Rouge, Malvern  43.806686   
1      M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2      M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3      M1G  Scarborough                                  Woburn  43.770992   
4      M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  Cluster Labels 1st Most Common Venue  \
0 -79.194353               0  Fast Food Restaurant   
1 -79.160497               0                   Bar   
2 -79.188711               0          Intersection   
3 -79.216917               0           Coffee Shop   
4 -79.239476               0           Gas Station   

        2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                 Yoga Studio            Donut Shop          Dessert Shop   
1  Construction & Landscaping           Yoga Studio            Donut Shop   
2           Electronics Store   Rental Car Location           Pizza Place   
3           Korean Restaurant           Yoga Studio      Doner Restaurant   
4                      Bakery   Fried Chicken Joint  Caribbean Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0    Dim Sum Restaurant                 Diner        Discount Store   
1    Dim Sum Restaurant                 Diner        Discount Store   
2        Breakfast Spot        Medical Center    Mexican Restaurant   
3          Dessert Shop    Dim Sum Restaurant                 Diner   
4       Thai Restaurant    Athletics & Sports                  Bank   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0               Dog Run      Doner Restaurant              Drugstore  
1               Dog Run      Doner Restaurant              Drugstore  
2           Yoga Studio          Dessert Shop     Dim Sum Restaurant  
3        Discount Store               Dog Run              Drugstore  
4      Hakka Restaurant             Drugstore             Donut Shop

In [30]:
Toronto_merged.shape

(102, 16)

## Map the clusters

In [31]:
# create map
map_clusters = folium.Map(location=[43.660719, -79.376603], zoom_start=10)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(Toronto_merged['Latitude'], Toronto_merged['Longitude'], Toronto_merged['Neighbourhood'], Toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### Toronto seems to be homogeneous in terms of venues category as one cluster includes most of the Neighbourhood. 

## Cluster 1  - 84 Neighbourhoods among 103 - Food & services

In [32]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 0, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels      1st Most Common Venue  \
0         Scarborough               0       Fast Food Restaurant   
1         Scarborough               0                        Bar   
2         Scarborough               0               Intersection   
3         Scarborough               0                Coffee Shop   
4         Scarborough               0                Gas Station   
6         Scarborough               0                Bus Station   
7         Scarborough               0                     Bakery   
8         Scarborough               0        American Restaurant   
9         Scarborough               0            College Stadium   
10        Scarborough               0          Indian Restaurant   
11        Scarborough               0  Middle Eastern Restaurant   
12        Scarborough               0  Latin American Restaurant   
13        Scarborough               0                Pizza Place   
15        Scarborough               0       Fast Food Restaurant   
16         North York               0   Mediterranean Restaurant   
17         North York               0             Clothing Store   
18         North York               0                       Café   
19         North York               0                  Cafeteria   
20         North York               0                  Piano Bar   
21         North York               0           Ramen Restaurant   
23         North York               0                   Pharmacy   
25         North York               0        Japanese Restaurant   
26         North York               0                Coffee Shop   
27         North York               0                Coffee Shop   
28         North York               0         Falafel Restaurant   
30         North York               0              Grocery Store   
32         North York               0              Grocery Store   
33         North York               0      Portuguese Restaurant   
34          East York               0                Pizza Place   
35          East York               0                   Pharmacy   
..                ...             ...                        ...   
66   Downtown Toronto               0      Vietnamese Restaurant   
67   Downtown Toronto               0            Airport Service   
68   Downtown Toronto               0                Coffee Shop   
69   Downtown Toronto               0                Coffee Shop   
70         North York               0             Clothing Store   
71         North York               0        Japanese Restaurant   
72               York               0                 Playground   
74   Downtown Toronto               0              Grocery Store   
75       West Toronto               0                     Bakery   
76       West Toronto               0                        Bar   
77       West Toronto               0                Coffee Shop   
79               York               0         Turkish Restaurant   
80               York               0              Grocery Store   
81       West Toronto               0         Mexican Restaurant   
82       West Toronto               0                  Gift Shop   
83       West Toronto               0                Pizza Place   
84   Downtown Toronto               0                Coffee Shop   
85       Queen's Park               0                Coffee Shop   
86        Mississauga               0                      Hotel   
87       East Toronto               0                        Spa   
88          Etobicoke               0                Coffee Shop   
89          Etobicoke               0                Pizza Place   
92          Etobicoke               0              Grocery Store   
93          Etobicoke               0                Golf Course   
94          Etobicoke               0                       Café   
95         North York               0        Empanada Restaurant   
98          Etobicoke               0             Discount Store   
99

## Cluster 2 - 2 Neighbourhoods among 103 - Shops

In [33]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 1, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
5       Scarborough               1            Playground   
47  Central Toronto               1                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
5            Yoga Studio      Doner Restaurant      Department Store   
47            Playground           Yoga Studio      Doner Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
5           Dessert Shop    Dim Sum Restaurant                 Diner   
47      Department Store          Dessert Shop    Dim Sum Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
5         Discount Store               Dog Run             Donut Shop  
47                 Diner        Discount Store                Dog Run

## Cluster 3 - 13 Neighbourhoods among 103 - Park & Sports

In [34]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 2, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
14       Scarborough               2                  Park   
22        North York               2     Convenience Store   
24        North York               2                  Park   
29        North York               2                  Park   
39         East York               2     Convenience Store   
43   Central Toronto               2                  Park   
49  Downtown Toronto               2                  Park   
63   Central Toronto               2                  Park   
73              York               2         Women's Store   
78        North York               2                  Park   
90         Etobicoke               2                 River   
91         Etobicoke               2                  Park   
97              York               2                  Park   

   2nd Most Common Venue       3rd Most Common Venue 4th Most Common Venue  \
14            Playground            Sculpture Garden           Yoga Studio   
22                  Park                        Bank           Flower Shop   
24     Food & Drink Shop                 Yoga Studio      Doner Restaurant   
29               Airport                 Snack Place           Yoga Studio   
39                  Park                 Yoga Studio          Dessert Shop   
43           Swim School                    Bus Line           Yoga Studio   
49            Playground                       Trail           Yoga Studio   
63      Sushi Restaurant               Jewelry Store                 Trail   
73                  Park        Fast Food Restaurant                Market   
78                Bakery  Construction & Landscaping      Basketball Court   
90                  Park                        Pool        Discount Store   
91        Baseball Field                 Yoga Studio      Doner Restaurant   
97           Yoga Studio            Doner Restaurant      Department Store   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
14        Discount Store         Deli / Bodega      Department Store   
22      Doner Restaurant    Dim Sum Restaurant                 Diner   
24          Dessert Shop    Dim Sum Restaurant                 Diner   
29      Doner Restaurant          Dessert Shop    Dim Sum Restaurant   
39    Dim Sum Restaurant                 Diner        Discount Store   
43               Dog Run          Dessert Shop    Dim Sum Restaurant   
49        Discount Store         Deli / Bodega      Department Store   
63               Dog Run          Dessert Shop    Dim Sum Restaurant   
73      Doner Restaurant          Dessert Shop    Dim Sum Restaurant   
78           Yoga Studio            Donut Shop                 Diner   
90          Dance Studio         Deli / Bodega      Department Store   
91          Dessert Shop    Dim Sum Restaurant                 Diner   
97          Dessert Shop    Dim Sum Restaurant                 Diner   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
14          Dessert Shop    Dim Sum Restaurant                  Diner  
22        Discount Store               Dog Run            Yoga Studio  
24        Discount Store               Dog Run             Donut Shop  
29                 Diner        Discount Store                Dog Run  
39               Dog Run      Doner Restaurant             Donut Shop  
43                 Diner        Discount Store       Doner Restaurant  
49          Dessert Shop    Dim Sum Restaurant                  Diner  
63                 Diner        Discount Store       Doner Restaurant  
73                 Diner        Discount Store                Dog Run  
78        Discount Store               Dog Run       Doner Restaurant  
90          Dessert Shop    Dim Sum Restaurant                  Diner  
91        Discount Store               Dog Run             Donut Shop  
97        Discount Store               Dog Run             Donut Shop

## Cluster 4 - 1 Neighbourhood 

In [35]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 3, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
62  Central Toronto               3                Garden   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
62           Yoga Studio      Doner Restaurant          Dessert Shop   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
62    Dim Sum Restaurant                 Diner        Discount Store   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
62               Dog Run            Donut Shop          Deli / Bodega

## Cluster 5 - 2 Neighbourhoods - diverse venues

In [36]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 4, Toronto_merged.columns[[1 ]+ list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
31  North York               4            Food Truck        Baseball Field   
96  North York               4        Baseball Field           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
31           Yoga Studio      Doner Restaurant    Dim Sum Restaurant   
96      Doner Restaurant          Dessert Shop    Dim Sum Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
31                 Diner        Discount Store               Dog Run   
96                 Diner        Discount Store               Dog Run   

   9th Most Common Venue 10th Most Common Venue  
31            Donut Shop       Department Store  
96            Donut Shop     Falafel Restaurant